pkg/supervisor/supervisor.go

```diff
@@ -0,0 +1,247 @@
+// including workers.
+//
+// The graceful shutdown sequence shuts down workers in an order that
+// respects worker dependencies.
+func (s *Supervisor) Run() []error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for len(s.workers) > 0 {
+		s.reconcile()
+		s.changed.Wait()
+	}
+	return s.errors
+}
+
+// Triggers a graceful shutdown sequence. This can be invoked from any
+// goroutine.
+func (s *Supervisor) Shutdown() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.shuttingDown = true
+	s.changed.Broadcast()
+}
+
+func (s *Supervisor) dependents(worker *Worker) (result []*Worker) {
+	for _, n := range s.names {
+		w := s.workers[n]
+		for _, r := range w.Requires {
+			if r == worker.Name {
+				result = append(result, w)
+				break
+			}
+		}
+	}
+	return
+}
+
+// make sure anything that would like to be running is actually
+// running
+func (s *Supervisor) reconcile() {
+	if s.shuttingDown {
+		s.reconcileShutdown()
+	} else {
+		s.reconcileNormal()
+	}
+}
+
+// shutdown anything that is safe to shutdown and not already shutdown
+func (s *Supervisor) reconcileShutdown() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process != nil && !w.process.shutdownClosed {
+			for _, d := range s.dependents(w) {
+				if s.workers[d.Name].process != nil {
+					log.Printf("cannot shutdown %s, %s still running", n, d.Name)
+					continue OUTER
+				}
+			}
+			close(w.process.shutdown)
+			w.process.shutdownClosed = true
+		}
+	}
+}
+
+// launch anything that is safe to launch and not already launched
+func (s *Supervisor) reconcileNormal() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process == nil {
+			for _, r := range w.Requires {
+				process := s.workers[r].process
+				if process == nil || !process.ready {
+					log.Printf("cannot start %s, %s not ready", n, r)
+					continue OUTER
+				}
+			}
+			s.launch(w)
+		}
+	}
+}
+
+func (s *Supervisor) launch(worker *Worker) {
+	context, cancel := context.WithCancel(s.context)
+	process := &Process{
+		Supervisor: s,
+		Worker:     worker,
+		Context:    context,
+		cancel:     cancel,
+		shutdown:   make(chan struct{}),
+	}
+	worker.process = process
+	go func() {
+		var err error
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					err = errors.Errorf("PANIC: %v", r)
+				}
+			}()
+			err = worker.Work(process)
+		}()
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+		worker.process = nil
+		if err != nil {
+			process.Log(err)
+			if worker.Retry {
+				if s.shuttingDown {
+					s.remove(worker)
+				} else {
+					process.Log("retrying...")
+				}
+			} else {
+				s.remove(worker)
+				s.errors = append(s.errors, err)
+				s.shuttingDown = true
+			}
+		} else {
+			s.remove(worker)
+		}
+		s.changed.Broadcast()
+	}()
+}
+
+type Process struct {
+	Supervisor *Supervisor
+	Worker     *Worker
+	// Used for hard cancel.
+	Context context.Context
+	cancel  context.CancelFunc
+	// Used to signal graceful shutdown.
+	shutdown       chan struct{}
```

Comment:
The asymmetry of having to select on `process.Shutdown()` for graceful-shutdown and `process.Context.Done()` for not-so-graceful-shutdown seems weird.  Why not use a context for both? Nothing about either of those names suggests which one is graceful and which one is not-so-graceful.

Reply:
This is the last outstanding comment... I agree the names could be better... my rationale for not using a context is that in my head shutdown is not actually a signal to end anything, it's a signal to begin something (i.e. the shutdown process) which could take an indefinite amount of time. This struck me as a bit of a departure from the context use case. This is sorta in line with what the http.Server type does in the std library (they have Server.Shutdown(context)).

How strongly do you feel about this? I have a minor preference for keeping the asymmetry but trying to come up with better naming and adding docs. If you feel strongly or are aware of some similar uses of contexts I'm happy to defer to your judgement.

Reply:
I don't feel super strongly about this.  But I will note that `http.Server` (and much of the standard library) is a little non-idiomatic in its use of contexts, because contexts were a retrofit in go1.7.